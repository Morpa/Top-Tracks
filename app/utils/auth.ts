import dotenv from 'dotenv'

dotenv.config()

const state = {
  token: null,
  expires: 0
}

export const getToken = async () => {
  const now = Math.round(new Date().getTime() / 1000)
  const remaining = state.expires - now

  if (remaining <= 60) {
    state.token = null
    state.expires = 0
  }

  if (state.expires === 0) {
    const { access_token, expires_in } = await fetch(
      'https://accounts.spotify.com/api/token',
      {
        method: 'POST',
        headers: {
          Authorization: `Basic ${Buffer.from(
            process.env.SPOTIFY_CLIENT_ID +
              ':' +
              process.env.SPOTIFY_CLIENT_SECRET
          ).toString('base64')}`,
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: 'grant_type=client_credentials'
      }
    ).then((response) => response.json())

    state.token = access_token
    state.expires = now + expires_in
  }

  return state.token
}
