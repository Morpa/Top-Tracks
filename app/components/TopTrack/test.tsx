import { render, screen } from '~/utils/test-utils'
import { TopTrack } from '.'

const props = {
  title: "Just Can't Get Enough - 2006 Remaster",
  image_url: 'https://i.scdn.co/image/ab67616d0000b27314ed850dd8267a90b436ac5d',
  spotify_url: 'https://open.spotify.com/track/0xuXRQJfxmdP6gvfDe2Z4L'
}

describe('<TopTrack />', () => {
  it('should render the heading', () => {
    const { container } = render(<TopTrack {...props} />)

    expect(
      screen.getByRole('img', {
        name: /just can't get enough - 2006 remaster/i
      })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a link', () => {
    render(<TopTrack {...props} />)

    expect(
      screen
        .getByRole('link', {
          name: /just can't get enough - 2006 remaster/i
        })
        .closest('a')
    ).toHaveAttribute('href', `${props.spotify_url}`)
  })
})
