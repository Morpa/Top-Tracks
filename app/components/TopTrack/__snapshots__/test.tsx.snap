// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopTrack /> should render the heading 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #FAFAFA;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 5.6rem;
  position: relative;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  color: #FAFAFA;
}

.c3 svg {
  width: 4rem;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 2.4rem;
}

.c2 {
  border-radius: 50%;
  position: absolute;
  object-fit: cover;
  width: 12rem;
  height: 12rem;
}

.c5 {
  color: #FAFAFA;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 2.8rem;
  font-weight: 600;
  margin-left: 10rem;
}

@media (max-width:768px) {
  .c0 {
    padding: 1.6rem;
  }
}

@media (max-width:768px) {
  .c2 {
    width: 6rem;
    height: 6rem;
  }
}

@media (max-width:768px) {
  .c5 {
    margin-left: 4.0rem;
    font-size: 1.2rem;
  }
}

<li
  class="c0"
  data-cy="top-track"
>
  <div
    class="c1"
  >
    <img
      alt="Just Can't Get Enough - 2006 Remaster"
      class="c2"
      src="https://i.scdn.co/image/ab67616d0000b27314ed850dd8267a90b436ac5d"
    />
    <div
      class="c3"
    >
      <svg
        aria-hidden="true"
        aria-label="Play Button"
        class="c4"
        data-cy="play"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M133 440a35.37 35.37 0 01-17.5-4.67c-12-6.8-19.46-20-19.46-34.33V111c0-14.37 7.46-27.53 19.46-34.33a35.13 35.13 0 0135.77.45l247.85 148.36a36 36 0 010 61l-247.89 148.4A35.5 35.5 0 01133 440z"
        />
      </svg>
    </div>
  </div>
  <a
    class="c5"
    href="https://open.spotify.com/track/0xuXRQJfxmdP6gvfDe2Z4L"
  >
    Just Can't Get Enough - 2006 Remaster
  </a>
</li>
`;
